Add option to make Post Featured Image a link

ntsekouras opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
This PR adds an option to make Post Featured Image block a link. ~It handles `target` and `rel` attributes as well.~
<!-- Please describe what you have changed or added -->

## How has this been tested?
Locally

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ntsekouras committed:
Add option to link Post Featured Image

aristath commented on packages/block-library/src/post-featured-image/block.json:
```diff
@@ -1,6 +1,21 @@
+		},
+		"linkTarget": {
+			"type": "string",
+			"default": "_blank"
```
Why are we defaulting to `_blank`?

ntsekouras replied:
Hm.. I guess it was natural to me as I've been expecting and prefer that approach, but after your comment and some research seems to be incorrect. Thanks for pointing this out Ari.

I have pushed the changes to default and changed PostTitle default as well.

ntsekouras committed:
make default to open in same tabs

aristath reviewed (approved):
Thank you @ntsekouras! LGTM :+1:

jasmussen commented:
Nice PR. I was about to echo Ari that defaulting to same window is always best :) 

This is what I see:

<img width="1059" alt="Screenshot 2020-09-30 at 09 20 16" src="https://user-images.githubusercontent.com/1204802/94655030-3d665f00-02fe-11eb-964e-bd712e68199f.png">

This is all in the json configuration, correct? No UI?

ntsekouras committed:
show chip in both cases for placeholder

ntsekouras commented:
Hey @jasmussen!

>This is all in the json configuration, correct? No UI?

Actually no. There are two cases where a placeholder is shown.
1. When there is no query context (that's your case in default front-page with show latest posts).
2. When there is context but no Featured Image is set.

I pushed a new. commit to show in both cases the `chip` we've been discussing in the other PR. Can you take it for a spin again? :) 

You can test this with Query block as well.

jasmussen commented:
Nice, this is closer. I see this when no featured image:

<img width="1164" alt="Screenshot 2020-09-30 at 10 26 04" src="https://user-images.githubusercontent.com/1204802/94661664-a56d7300-0307-11eb-9a4a-5960c614dac0.png">

And this when I do:

![link](https://user-images.githubusercontent.com/1204802/94661683-ad2d1780-0307-11eb-8ce8-38c87dd4fd82.gif)

The link UI is rather different than the Image link UI, though:

![link](https://user-images.githubusercontent.com/1204802/94661829-dc438900-0307-11eb-9011-cb39b37577ed.gif)

This presents a few questions:

- What does the featured image link to? The post? The image itself? Can I write my own custom link?
- Since the featured image is entirely a template part, you should be able to edit and configure it even without the right content. I.e. I need to be able to configure that a featured image links, without having to write a post with a featured image first. 
- If the link indeed does link to the post itself the toggle should be rephrased to say "Link to post". (or post type?). And also — should a link to a post ever open in a new tab, and what would be the use case of the `rel`?

aristath commented:
Should it even have the "Open in new link" option? :thinking: 
Since the image will always link to the same thing and is not editable, I don't think we'll ever need to open in a new tab... :thinking:

ntsekouras commented:
Great questions!

>What does the featured image link to? The post? The image itself? Can I write my own custom link?

It links to the entity and not the image. I think it might be considered to link to the image in a follow up PR. It doesn't make much sense to me to allow a custom link as this can be achieved with the `Image` block and `PostFeaturedImage` is much more specific to the current post.

>Since the featured image is entirely a template part, you should be able to edit and configure it even without the right content. I.e. I need to be able to configure that a featured image links, without having to write a post with a featured image first.

That makes sense yes. Especially in Query block usage.

>If the link indeed does link to the post itself the toggle should be rephrased to say "Link to post". (or post type?). And also — should a link to a post ever open in a new tab, and what would be the use case of the rel?

>Should it even have the "Open in new link" option? Since the image will always link to the same thing and is not editable, I don't think we'll ever need to open in a new tab..

The label could become more specific like `Make featured image a link to {entity}`. Regarding `new tab` and `rel`, I think it's up for the user to decide if wants to enable and what `rel` to add there..

I'll make the above changes.

jasmussen commented:
To be clear, what I'm advocating for is to *start with fewer features*. It's always easier to add features once it becomes evident that they are necessary, whereas it is often hard to _remove_ features when they turn out to be mostly unnecessary.

In that vein, I would recommend:

- Link toggle working for both placeholder and empty state
- No external link toggle, no rel feature until we know we need it
- Simpler phrasing of the toggle.

> The label could become more specific like `Make featured image a link to {entity}`. 

Nice. Can we shorten it and be more action'y? Such as: `Link to {entity}`? Keep in mind, there's this right above providing ample context:

<img width="282" alt="Screenshot 2020-09-30 at 10 57 16" src="https://user-images.githubusercontent.com/1204802/94664873-bb7d3280-030b-11eb-9002-31e2e6ff128e.png">

ntsekouras committed:
address review feedback

ntsekouras commented:
The changes I made are:

1. remove `target` and `rel`.
2. Changed the label to also mention the post type of the context it lives.
3. Made the option show even if no featured image is set. This doesn't apply when there is no context like I mentioned above. I'm not sure how to handle this yet, but I think it's related with a decision that affects all PostBlocks.
4. Removed the `link` entirely from `edit` as it visually changes nothing and it makes it hard to select the block, if the image is wide.

ntsekouras committed:
fix whitespace
